src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -47,405 +49,429 @@
+        grpc_transport* transport = grpc_create_chttp2_transport(
+            args->args, args->endpoint, false, resource_user);
+        grpc_server_setup_transport(
+            self->listener_->server_, transport, self->accepting_pollset_,
+            args->args, grpc_chttp2_transport_get_socket_node(transport),
+            resource_user);
+        // Use notify_on_receive_settings callback to enforce the
+        // handshake deadline.
+        self->transport_ = reinterpret_cast<grpc_chttp2_transport*>(transport);
+        self->Ref().release();  // Held by OnReceiveSettings().
+        GRPC_CLOSURE_INIT(&self->on_receive_settings_, OnReceiveSettings, self,
+                          grpc_schedule_on_exec_ctx);
+        grpc_chttp2_transport_start_reading(transport, args->read_buffer,
+                                            &self->on_receive_settings_);
+        grpc_channel_args_destroy(args->args);
+        self->Ref().release();  // Held by OnTimeout().
+        GRPC_CHTTP2_REF_TRANSPORT((grpc_chttp2_transport*)transport,
+                                  "receive settings timeout");
+        GRPC_CLOSURE_INIT(&self->on_timeout_, OnTimeout, self,
+                          grpc_schedule_on_exec_ctx);
+        grpc_timer_init(&self->timer_, self->deadline_, &self->on_timeout_);
+      } else {
+        if (resource_user != nullptr) {
+          grpc_resource_user_free(resource_user,
+                                  GRPC_RESOURCE_QUOTA_CHANNEL_SIZE);
+        }
+      }
     }
+    self->handshake_mgr_->RemoveFromPendingMgrList(
+        &self->listener_->pending_handshake_mgrs_);
   }
-  connection_state->handshake_mgr->RemoveFromPendingMgrList(
-      &connection_state->svr_state->pending_handshake_mgrs);
-  gpr_mu_unlock(&connection_state->svr_state->mu);
-  connection_state->handshake_mgr.reset();
-  gpr_free(connection_state->acceptor);
-  grpc_tcp_server_unref(connection_state->svr_state->tcp_server);
-  server_connection_state_unref(connection_state);
+  self->handshake_mgr_.reset();
+  gpr_free(self->acceptor_);
+  grpc_tcp_server_unref(self->listener_->tcp_server_);
+  self->Unref();
 }
 
-static void on_accept(void* arg, grpc_endpoint* tcp,
-                      grpc_pollset* accepting_pollset,
-                      grpc_tcp_server_acceptor* acceptor) {
-  server_state* state = static_cast<server_state*>(arg);
-  gpr_mu_lock(&state->mu);
-  if (state->shutdown) {
-    gpr_mu_unlock(&state->mu);
-    grpc_endpoint_shutdown(tcp, GRPC_ERROR_NONE);
-    grpc_endpoint_destroy(tcp);
-    gpr_free(acceptor);
-    return;
+//
+// Chttp2ServerListener
+//
+
+grpc_error* Chttp2ServerListener::Create(grpc_server* server, const char* addr,
+                                         grpc_channel_args* args,
+                                         int* port_num) {
+  std::vector<grpc_error*> error_list;
+  grpc_resolved_addresses* resolved = nullptr;
+  Chttp2ServerListener* listener = nullptr;
+  grpc_error* error = [&]() {
```

Comment:
This is pretty large as a lambda, especially since it's just called with an empty argument; would this make more sense as a helper function?

Reply:
The only reason I moved this into a lambda is to eliminate the use of C-style goto.  The logic inside of the lambda is pretty tightly coupled to this method; it I made it a helper function, it would require access to literally every local variable defined here, so it doesn't seem worth doing that.

Honestly, the real cause of the complexity here is that `grpc_tcp_server_create()` requires passing in the cleanup closure at construction time.  Since that closure is part of the listener, that means we need to allocate the listener before we know if the operation succeeds, which complicates the error-handling logic.  If it weren't for that, we could just use normal RAII techniques here.

For now, I've added a comment explaining why this lambda is here.